Refactor and optimize terrain generation

DaTa- opened the pull request:
TL;DR: attempt to reduce lag, stuttering and allow more players to catch up during heavy games. 

### Motivation
vscode profiler graph (how to setup https://www.youtube.com/watch?v=oNfMNFxy2X4 )
![old_profiler_run](https://github.com/user-attachments/assets/35e4e771-bc06-469b-8ce8-cc1e869c8c0e)
This graph was taken after map was fully charted and new chunks were generated by walking around silo

### Proof of improvement
Sections of code were wrapped with `LuaProfiler` and sane amount of measurement was gathered to guide optimization (all revealed map + 32 * 32 chunks square in the middle). Below chunk generation time statistics. Top row corresponds to all measurements and other rows correspond to chunk locations around the map. Time is in milliseconds, format is "min/average/max (median)".
Measurement was conducted on revision where chunk generation profiling was first introduced in code (rev. e96f3cdd), but no optimizations were applied yet:
```
all: 0.002/8.593/51.783 (5.862)
river: 0.543/7.620/17.186 (7.210)
ordinary: 0.002/5.817/15.175 (5.538)
biter_area_border: 3.644/23.327/51.783 (23.378)
biter_area: 5.520/10.471/24.717 (9.642)
```
And this is a measurement of the last commit:
```
all: 0.022/1.956/10.765 (1.371)
river: 0.646/2.259/4.919 (2.197)
ordinary: 0.022/1.005/9.316 (0.132)
biter_area_border: 0.133/4.871/10.765 (5.048)
biter_area: 2.862/4.432/8.726 (4.349)
```

And for mixed ore map special (mixed patches, default size 4), before:
```
all: 1.563/8.337/64.389 (5.822)
river: 3.039/8.578/16.518 (8.103)
ordinary: 1.563/6.320/18.118 (5.752)
biter_area_border: 4.120/23.355/64.389 (24.217)
biter_area: 2.832/5.254/18.581 (4.629)
```
After:
```
all: 0.038/2.508/13.577 (1.826)
river: 0.994/2.148/5.253 (1.919)
ordinary: 0.038/1.937/9.420 (1.597)
biter_area_border: 0.272/4.755/13.577 (4.788)
biter_area: 2.525/3.700/7.962 (3.651)
```

### Other optimizations
#### Terrain mirroring
Before optimizations:
```
invert_tiles: 1.597/2.604/26.804 (2.237)
```
After:
```
invert_tiles: 1.397/2.144/21.341 (1.879)
```

#### Spawn population (excluding chunk generation) (10 measures, same seed)
Before:
```
all: 1352.785/1364.642/1393.854 (1366.072)
draw_spawn_area: 1063.808/1072.760/1104.954 (1066.324)
clear_ore_in_main: 10.209/10.525/10.844 (10.636)
generate_spawn_ore: 271.322/272.995/280.089 (272.592)
generate_additional_rocks: 0.385/0.413/0.494 (0.410)
generate_silo: 5.944/7.558/17.172 (6.348)
draw_spawn_island: 0.249/0.274/0.414 (0.258)
```
After:
```
all: 535.826/554.246/569.986 (557.267)
draw_spawn_area: 475.224/489.776/509.943 (493.149)
clear_ore_in_main: 8.699/11.790/23.424 (9.099)
generate_spawn_ore: 42.512/44.235/47.106 (44.224)
generate_additional_rocks: 0.304/0.340/0.446 (0.327)
generate_silo: 6.141/7.715/19.775 (6.408)
draw_spawn_island: 0.239/0.263/0.376 (0.249)
```

### Bugfixes
- Fixed fish spawned inside land
- `storage.random_generator` was removed since it didn't provide determinism
- Some determinism fixes for ordinary and mixed ore map special

### Other
- Terrain generation was heavy refactored as the only way to introduce some beneficial optimizations
- Noise related functionality was cleaned up, refactored and gathered in one place
- Introduced module for with a hacky way of gathering profiling of multiple measurements (`profiler_stats.lua`), since no other alternative I have found
- Chunk generation profiling was left in code (`terrain.lua`, disabled)
- Mixed ores man special UI input was sanitized

### What was broken
Old seeds will give you same mixed ore patches, mostly same walls/turrets/chests, but duo to change of order and amount of rng invocation ore and silo layout couldn't be preserved. Also old way of random tree removal was more expensive then necessary, but despite my attempts to preserve similar amounts of trees at spawn, it seems removes more trees with bigger radius, so maybe someone wants to tinker with radius/probability values if needed
Old:
![Screenshot 2025-01-27 042722](https://github.com/user-attachments/assets/982c9610-572d-44c7-842b-991b3b9b1c6e)

New:
![Screenshot 2025-01-27 035427](https://github.com/user-attachments/assets/3f03e0f0-02bf-4b5d-ba6d-74180a51c38f)

### Testing
Unfortunately, only single player/2 players testing was performed, not enough to claim it as stable solution due to amount of changed code.

#### "Combat log" is available:
[chunk_profile.txt](https://github.com/user-attachments/files/18553189/chunk_profile.txt)

TheBigZet commented:
5x lag spike improvement? Sign me in!

RedRafe commented:
Thanks a lot @DaTa-, amazing work! I will take time to also review the changes, but I've already looked around and everything seems very good.

Terrain gen changes patch position - I'm not particularly worried about this, but I will ask @clifffrey if it is possible to package the current scenario as a new GitHub release, in order to maintain compatibility with all the "good seeds" that have been saved on Discord, since those would only be re-playable in the current version pre-optimization.

RedRafe commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -372,162 +650,351 @@ local function generate_extra_worm_turrets(surface, left_top)
+    return biter_area_noise_test(x, y, seed, a)
+end
+
+local spitter_spawner_template = { name = 'spitter-spawner', position = { 0, 0 }, force = 'north_biters' }
+local biter_spawner_template = { name = 'biter-spawner', position = { 0, 0 }, force = 'north_biters' }
+local worm_turret_template = { name = '', position = { 0, 0 }, force = 'north_biters' }
```
Check for desyncs

DaTa- replied:
`spitter_spawner_template` only position is modified before usage

`biter_spawner_template` only position is modified before usage

`worm_turret_template ` name and position are modified before 2 usages

DaTa- replied:
Moved inside function.

RedRafe commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -561,50 +1028,55 @@ function Public.draw_spawn_island(surface)
-            mixed_ore_map(surface, x * 32, y * 32)
-        end
-    end
+local water_barrier = {}
```
Check for desyncs

DaTa- replied:
Only `position[1]` (x coordinate) is modified for all line of 32 tiles before usage.

RedRafe commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -1035,30 +1481,30 @@ function Public.add_new_year_island_decorations(surface)
     end
 
-    for _ = 1, storage.random_generator(0, 6) do
+    for _ = 1, math_random(0, 6) do
```
Maybe the decoratives table can be populated all together before passing it to `surface.create_decoratives` for better perf, instead of creating and calling 1 at a time like entities (which doesnt have the collection method)

DaTa- replied:
I haven't touched new year island, only checked if it works after I removed `storage.random` generator

RedRafe commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -2,25 +2,90 @@ local Public = {}
+end
+
+-- avoid allocations to improve performance and maybe reduce gc lag
+local preallocated_out_of_map_tiles = {}
+local preallocated_tiles = {}
```
Check for desyncs

DaTa- replied:
This is a bit tricky. In theory, different clients gonna use different tiles from this array, but no more than all 1024 at the time. `next_preallocated_tile` value doesn't matter before chunk generation. They are being collected, `position`s are changed before usage and for `preallocated_tiles` `name` is also changed in both places of usage. But it's hard to prove definitely if it's 100% desync safe.

RedRafe reviewed (changes requested):
Mainly just worried about MP desyncs for some tables. I'll need to have a better look at `terrain.lua` because tehere's a lot of changes there, but everything else seems fine :)

DaTa- commented on maps/biter_battles_v2/functions.lua:
```diff
@@ -406,7 +364,7 @@ function Functions.no_turret_creep(event)
         posEntity.y = posEntity.y - 100
     end
-    if not Functions.is_biter_area(posEntity, false) then
+    if not Functions.is_roughly_biter_area(posEntity) then
```
Hey, thanks for the review!
`Functions.no_turret_creep` is outside of the scope of this PR. I didn't modify how it works, only decoupled 2 functionalities which `Functions.is_biter_area` provided in order to move it to `terrain.lua` and move noise related stuff out. If you inspect old version with `false` as `noise_Enabled` argument value, you can observe it works almost identically and it rarely called only when someone tries to place a turret.

DaTa- commented on maps/biter_battles_v2/mirror_terrain.lua:
```diff
@@ -124,32 +125,30 @@ function Public.remove_hidden_tiles(event)
     end
 end
 
+local tiles = {}
```
It contains exactly 1024 tiles, which all are overridden during whole chunk inversion, both `position` and `name` values are set before using it.

DaTa- commented on maps/biter_battles_v2/mixed_ore_map_special.lua:
```diff
@@ -0,0 +1,290 @@
+    mixed_ore_weight[k] = mixed_ore_weight_total
+end
+
+local ore_template = { name = 'iron-ore', position = { 0, 0 }, amount = 1 }
+local ore_template_no_tree_removal = { name = 'iron-ore', position = { 0, 0 }, amount = 1, enable_tree_removal = false }
```
This is a pre-allocated set of parameters for function calls. Optimization avoids reallocating them over and over for possibly 1024 times per chunk. They are being used in 2 cases:
1. Check for collision
2. Create ore entity

When checking for collision, only modification `position` parameter is necessary. `amount` and `enable_tree_removal` are not used in this situation. As for `name` field, it's a bit different: you can use what-ever ore name for collision check, because they all have same collision box and collision mask: https://github.com/wube/factorio-data/blob/3592a6af4140611ed34c4148066a2a7004d736c3/base/prototypes/entity/resources.lua (unset = default)

For creating ore all 3 fields must be overridden (excluding `enable_tree_removal`, which is unmodified at all times).

I carefully examined all places where I'm using these templates and actually found 1 not well tested place where I've misused it, so I'll push fixed version in a moment

RedRafe replied:
This is not an optimization issue, but a desync issue.
I recommend reading [Gangsir's modding tutorial multiplayer and desyncs section](https://wiki.factorio.com/Tutorial:Modding_tutorial/Gangsir#Multiplayer_and_desyncs) which can provide a better explanation than what I can provide

DaTa- commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -82,102 +151,132 @@ local loot_blacklist = {
-
-    local position = { x = tile.position.x, y = tile.position.y }
+---@enum area_intersection
+area_intersection = {
+    none = 1,
+    partial = 2,
+    full = 3,
+}
```
This is more like enum, not constants. It's never being used in performance critical spots, so should be fine?

DaTa- commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -82,102 +151,132 @@ local loot_blacklist = {
-    for _ = 1, count, 1 do
-        local tile_placed = false
+---@enum chunk_type
+chunk_type = {
+    river = 1,
+    ordinary = 2,
+    biter_area_border = 3,
+    biter_area = 4,
+}
```
Same thing again, `chunk_type_at`, which returns, it and further usages are not in performance critical spots (can't prove it improved performance with profiler)

DaTa- commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -82,102 +151,132 @@ local loot_blacklist = {
-    end
-    local seed = game.surfaces[storage.bb_surface_name].map_gen_settings.seed
-    local noise_seed_add = 25000
+local remove_structures_template = { position = { 0, 0 }, name = { 'wooden-chest', 'stone-wall', 'gun-turret' } }
+local ore_template = { name = 'iron-ore', position = { 0, 0 }, amount = 1 }
```
I checked `remove_structures_template`, only `position` is modified before 1 usage just after it

I checked `ore_template` too with alongside with `mixed_ore_map_special.lua`, see comment there

DaTa- replied:
Moved inside function (but not other `ore_template` global).

DaTa- commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -186,167 +285,340 @@ local function draw_noise_ore_patch(position, name, surface, radius, richness)
+local concrete_circle_radius = spawn_wall_radius - (10 + spawn_wall_noise_deviation)
+local spawn_wall_noise_radius = spawn_wall_radius + (4.5 + spawn_wall_noise_deviation)
+
+local wooden_chest_template = { name = 'wooden-chest', position = { 0, 0 }, force = 'north' }
+local coal_template = { name = 'coal', position = { 0, 0 } }
+local stone_wall_template = { name = 'stone-wall', position = { 0, 0 }, force = 'north' }
+local gun_turret_template = { name = 'gun-turret', position = { 0, 0 }, force = 'north' }
+local gun_turret_remnants_template = { name = 'gun-turret-remnants', position = { 0, 0 }, force = 'neutral' }
+local fire_magazine_template = { name = 'firearm-magazine', count = 0 }
```
Checked `wooden_chest_template`, `name` and `position` are modified before every usage.

`coal_template` - `position` modified before single usage.

`stone_wall_template ` - `position` modified before every usage.

`gun_turret_template ` - I've found 1 desyncing place where I've misused it, ill force push fixed version later. After fix: `position` is modified before usage.

`gun_turret_remnants_template` - (after fixed above) only position is modified before usage

`fire_magazine_template` - only count is modified before usage

DaTa- replied:
Moved inside function.

DaTa- commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -186,167 +285,340 @@ local function draw_noise_ore_patch(position, name, surface, radius, richness)
+            end
+
+            if distance_from_spawn_wall > -1.95 and distance_from_spawn_wall < 0 then
+                surface.create_entity({ name = 'stone-wall', position = pos, force = 'north' })
```
Good catch, I'll fix it

DaTa- commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -186,167 +285,340 @@ local function draw_noise_ore_patch(position, name, surface, radius, richness)
+    return chunk_pos.x < -5 or chunk_pos.x >= 5 or chunk_pos.y < -5
+end
+
+local ore_template = { name = 'iron-ore', position = { 0, 0 }, amount = 1 }
```
Checked in comment above

DaTa- commented:
I applied fixes/improvements suggested and tested new version for desyncs in multiplayer during generation

TheBigZet commented on utils/profiler_stats.lua:
```diff
@@ -0,0 +1,92 @@
+        ---@param profiler LuaProfiler used and stopped factorio profiler. Don't reuse this profiler after supplying it as the argument to this function
+        ---@return boolean
+        add_record = function(profiler)
+            local player = game.player
```
I'm 99% sure this will never be true. `game.player` exists only after direct interaction with game console
https://lua-api.factorio.com/latest/classes/LuaGameScript.html#player

DaTa- replied:
If left it in because if someone tries to run a command with profiling during multiplayer and they get some unexpected locale from other player, at least he in theory have control to fix it. In fact, I should fix other place which doesn't account for that

TheBigZet commented on utils/profiler_stats.lua:
```diff
@@ -0,0 +1,92 @@
+        ---@return boolean
+        add_record = function(profiler)
+            local player = game.player
+            if not player then
+                for _, p in pairs(game.players) do -- `next` doesn't work with `LuaCustomTable`
+                    player = p
+                    break
+                end
+                if not player then
+                    log('warning: as a workaround, at least 1 player is required for accessing factorio profiling')
+                    return false
+                end
+            end
```
Does any player need to be connected? Then how about `game.connected_players[1]`?
Doesn't need to be connected? Try `game.players[1]`
You can check number of those with `#`

DaTa- replied:
This is a player index, not an array index, documentation is too dull about it
![Screenshot 2025-01-27 214320](https://github.com/user-attachments/assets/fe3c7f75-bf85-40d5-a05b-1fe32fd2908a)

DaTa- replied:
Aw, I've missed it in the documentation

TheBigZet commented on utils/profiler_stats.lua:
```diff
@@ -0,0 +1,92 @@
+                end
+            end
+
+            -- In order to prevent desyncs, factorio disallows to directly read measures from `LuaProfiler`,
+            -- it expected to be directly printed without being observed by the game state. Fortunately,
+            -- we can leverage (abuse) `LuaPlayer::request_translation` to convert this lua object to
+            -- localized string.
```
So you want to do sth with those values other than printing?

DaTa- replied:
Correct.

clifffrey reviewed (commented):
Wow!  This is so awesome!  Thank you so much!

This is a huge change that looks like it was a lot of work to create, and like it will be a huge win for our players!

DaTa- commented on maps/biter_battles_v2/mirror_terrain.lua:
```diff
@@ -124,32 +125,30 @@ function Public.remove_hidden_tiles(event)
-                name = tile.name,
-            }
-        end
+    for i, src_tile in pairs(to_emplace) do
```
`clone_area` is called in exactly 1 place in the scenario, `defines.events.on_area_cloned` event is registered in exactly 1 place in the scenario.

`on_area_cloned` callback expects to be called only for this one call, and if it's not true, you are screwed anyway. It's the only place from which `invert_tiles` is called.

`invert_tiles` uses area supplied from chunk generation event (always 32x32 tiles, there's no "empty" tile could exist), mirrored.

When I was testing this change, I indeed placed assert there to check my assumption even further.

DaTa- commented on utils/profiler_stats.lua:
```diff
@@ -0,0 +1,92 @@
+    return result and tonumber(result)
+end
+
+-- circumventing desync protection
```
I'll update the commentary.

DaTa- commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -82,102 +151,132 @@ local loot_blacklist = {
+                if noise then
+                    local pos = { ore_x, ore_y }
+                    ore_template.position = pos
+                    if can_place_entity(ore_template) then
```
I already tried
![Screenshot 2025-01-29 014132](https://github.com/user-attachments/assets/d417400d-0609-4281-b8ad-0811a9d31679)

RedRafe replied:
If you'd like to drop your +1 on this issue maybe it's likely to get implemented and we can get another x2 boost

https://forums.factorio.com/viewtopic.php?t=126368

DaTa- commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -360,8 +633,14 @@ for x = -8, 8, 1 do
 
-local function generate_extra_worm_turrets(surface, left_top)
-    local chunk_distance_to_center = math_sqrt(left_top.x ^ 2 + left_top.y ^ 2)
+local worm_template = { name = '', position = { 0, 0 }, force = 'north_biters' }
```
Done.

DaTa- commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -186,167 +285,341 @@ local function draw_noise_ore_patch(position, name, surface, radius, richness)
+    end
+end
+
+local fish_template = { name = 'fish', position = { 0.5, 0.5 } }
```
Done

DaTa- commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -186,167 +285,341 @@ local function draw_noise_ore_patch(position, name, surface, radius, richness)
+local concrete_circle_radius = spawn_wall_radius - (10 + spawn_wall_noise_deviation)
+local spawn_wall_noise_radius = spawn_wall_radius + (4.5 + spawn_wall_noise_deviation)
+
+local wooden_chest_template = { name = 'wooden-chest', position = { 0, 0 }, force = 'north' }
```
Moved inside.

DaTa- commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -82,102 +151,132 @@ local loot_blacklist = {
-    end
-    local seed = game.surfaces[storage.bb_surface_name].map_gen_settings.seed
-    local noise_seed_add = 25000
+local remove_structures_template = { position = { 0, 0 }, name = { 'wooden-chest', 'stone-wall', 'gun-turret' } }
```
Moved inside.

DaTa- commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -372,162 +651,351 @@ local function generate_extra_worm_turrets(surface, left_top)
+    return biter_area_noise_test(x, y, seed, a)
+end
+
+local spitter_spawner_template = { name = 'spitter-spawner', position = { 0, 0 }, force = 'north_biters' }
```
Done.

DaTa- commented on maps/biter_battles_v2/mixed_ore_map_special.lua:
```diff
@@ -0,0 +1,290 @@
+    mixed_ore_weight[k] = mixed_ore_weight_total
+end
+
+local ore_template = { name = 'iron-ore', position = { 0, 0 }, amount = 1 }
```
I'm not gonna do it, it hurts performance.

DaTa- commented on utils/profiler_stats.lua:
```diff
@@ -0,0 +1,92 @@
+
+local Event = require('utils.event')
+
+local pub = {}
```
I will next time :)

DaTa- replied:
Done.

DaTa- commented:
I made some changes, mostly to reduce desync risks in the eyes of code reviewers. If you disagree with some leftover globals - feel free to enable chunk profiling in `terrain.lua`, turn off all noise producing services/apps on your pc, run 4-6 times profiling before your suggested change and 4-6 after it to prove it didn't harm performance.

clifffrey reviewed (commented):
due to force-push, I am unable to review the delta between what I reviewed last time and what is there now.

It would be helpful for me if you could push 4eb6a2abab19b386dae14e06118fee8f02f44645 to a branch somewhere so that I can actually compare what is there now to what I reviewed before.
In the future, it is helpful for reviewers if you avoid force-pushing.

DaTa- commented:
You were criticizing me for fragmenting my commits to much when I add them after review, now you criticizing me for squashing it, can you guys gather at #dev-git and discuss some coherent rules for submitting a pull request?

I don't fully understand what do you want from me, please make it more clear (maybe you gave wrong revision number too). You can compare changes here, if you meant this:
![image](https://github.com/user-attachments/assets/d69e603e-a4e6-4164-9f40-288dc0733896)

clifffrey commented:
> You were criticizing me for fragmenting my commits to much when I add them after review, now you criticizing me for squashing it, can you guys gather at #dev-git and discuss some coherent rules for submitting a pull request?
> 
> I don't fully understand what do you want from me, please make it more clear (maybe you gave wrong revision number too). You can compare changes here, if you meant this: ![image](https://private-user-images.githubusercontent.com/670602/407606706-d69e603e-a4e6-4164-9f40-288dc0733896.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg0NjkwOTksIm5iZiI6MTczODQ2ODc5OSwicGF0aCI6Ii82NzA2MDIvNDA3NjA2NzA2LWQ2OWU2MDNlLWE0ZTYtNDE2NC05ZjQwLTI4OGRjMDczMzg5Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwMlQwMzU5NTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MWFjNmMyNzNkYjcxZjlkNjc5ZmVkNTBmZjM2M2E5ZjA5MDEzMTI2ODM3ZjNkZDQ2YTdhYmE5NWJkN2FhZTBiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.11kLubcvEUf4YQUS0UBozcwEhY3oQUdL4TktHAfVqwk)

I never intended to criticize you. I thought that I was only making suggestions about things that would make things easier for me. If there is a way that I could word my feedback differently, please let me know - I truly do not want to offend you. I was frustrated because in the github code review UI, when I clicked the "review changes since you last reviewed this change", github gave me errors saying that the version that I reviewed was unavailable.

Thank you for pointing out the other UI for comparing versions - that is very helpful.  And thank you for reducing the risk of this change by reducing the number of global variables, and rerunning the benchmarks. I understand that that is a lot of work, and I appreciate it.

If you don't want to rename the classes to Public, I can modify your PR to add a commit that does the rename, or perhaps we can convince a different developer to cherry pick your changes and do the rename. Let me know what you would prefer. I'm excited to be able to approve this change and get it merged.

DaTa- commented:
I've renamed exposed tables to `Public`, squashed last commit with bunch of desync phobia related changes into appropriate commits and also added some more guards to future proof/ease reasoning about desyncs where I could (some off scene attempts resulted in performance regression and I didn't apply them in the end)

DaTa- commented on maps/biter_battles_v2/mixed_ore_map_special.lua:
```diff
@@ -0,0 +1,290 @@
+    end
+end
+
+local pub = {}
```
Done.

developer-8 commented:
![image](https://github.com/user-attachments/assets/c73b3648-4d0b-4f20-a583-d6f087764074)
strange empty patch in Mixed ore special
seed 123125

DaTa- commented:
>  strange empty patch in Mixed ore special. Type Mixed ore, size default seed 123125
I forgot about this issue, this is duo to lake under concrete that was replaced with filament tiles

DaTa- commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -2,25 +2,90 @@ local Public = {}
-local chunk = require('maps.biter_battles_v2.precomputed.chunk_container')
 
+local ai_targets_start_tracking = AiTargets.start_tracking
+local bb_config_bitera_area_distance = bb_config.bitera_area_distance
+local bb_config_biter_area_slope = bb_config.biter_area_slope
```
> It was possible to change variables from `bb_config` in runtime
This claim isn't true:
![image](https://github.com/user-attachments/assets/a34521a6-10cf-40a3-a73e-a14ac197ed83)

developer-8 replied:
yes you are right

DaTa- commented on maps/biter_battles_v2/init.lua:
```diff
@@ -307,16 +273,11 @@ function Public.tables()
-    end
     if storage.next_map_seed == nil or storage.next_map_seed < 341 then
         -- Seeds 1-341 inclusive are the same
         -- https://lua-api.factorio.com/latest/classes/LuaRandomGenerator.html#re_seed
```
It's not well documented, but setting map seed to these values have bad effect even in free play:
![image](https://github.com/user-attachments/assets/36e71b58-5667-42e8-b789-e46faffad6cc)

developer-8 replied:
ok. Asked because `re_seed` not used anymore

clifffrey commented on maps/biter_battles_v2/terrain.lua:
```diff
@@ -1040,9 +1040,33 @@ local function draw_spawn_island(surface)
+
+--- Create initial chunks to be populated with spawn objects
+---@param surface LuaSurface
+local function force_north_part_map_generation(surface)
```
personally, I would combine these into one function that took an argument of "north" or "south" and turned that into -1 or 1 or whatever for the limits... but, shrug, it is also ok as you have it.

DaTa- replied:
I'll give it a try tomorrow

Also, found a bug introduced by the last change: dummy character spawns after map reset, gonna fix it later.

DaTa- replied:
Done.

Couldn't reproduce the bug with multiple players online, maybe some players were trolling with editor (there is a check in code to prevent dummy character cloning), needs more testing.

developer-8 commented:
![image](https://github.com/user-attachments/assets/dd7091b2-c630-461c-91df-05e0ff33851d)
If the reveal map is off, spawners may not be generated

DaTa- commented:
@developer-8 damn, how are you this good at testing? Fixed :)

developer-8 commented:
![12333_2](https://github.com/user-attachments/assets/c538240b-bcb2-4783-8a3c-60c29886e4e1)
![12333_1](https://github.com/user-attachments/assets/8d9a8839-c9b8-438b-810c-9f8120862cf1)
starting ore not the same when reset map with same seed

DaTa- commented:
Thanks @developer-8 . I inspected this bug and turns out it even exists in old terrain generation, but chances to trigger it very low due to same amount of rng calls (once for each tile, which is unnecessary high number) with coincidentally same order. I temporary removed random tree removal (in practice few trees will be missing from spawn area) and submitted a bug report https://forums.factorio.com/viewtopic.php?t=126774

clifffrey commented:
@DaTa-  do you believe that this is ready to merge?

Also, I'd ask that you hold off on making any changes that are not bug-fixes to this PR, if we ever want it to be able to be merged.

clifffrey reviewed (approved):
@DaTa- confirms that this should be ready to merge.  Of course, there might be bugs/etc, nothing is without risk, etc.

I'm very excited for this performance win!

developer-8 commented:
@DaTa- Can you please redo the performance measurements from the description after all these changes. I'm curious how it ended up

DaTa- commented:
> @DaTa- Can you please redo the performance measurements from the description after all these changes. I'm curious how it ended up. If it doesn't too difficult for you

```
No rebase (perf/chunk-gen branch)

all: 0.024/2.004/13.449 (1.520)
river: 0.486/2.269/5.572 (1.999)
ordinary: 0.024/1.030/9.388 (0.133)
biter_area_border: 0.160/5.149/13.449 (5.338)
biter_area: 2.989/4.401/7.037 (4.322)

all: 0.023/1.852/12.854 (1.381)
river: 0.193/1.501/3.401 (1.353)
ordinary: 0.023/1.017/9.612 (0.124)
biter_area_border: 0.157/4.843/12.854 (4.817)
biter_area: 2.734/3.814/5.053 (3.777)

Rebase (master branch)

all: 0.024/2.247/13.184 (1.672)
river: 0.478/2.200/4.951 (2.132)
ordinary: 0.024/1.322/9.555 (0.138)
biter_area_border: 0.149/5.028/13.184 (5.224)
biter_area: 3.252/4.822/9.780 (4.800)

all: 0.023/2.159/13.239 (1.649)
river: 0.371/2.455/5.868 (2.380)
ordinary: 0.023/1.185/9.711 (0.140)
biter_area_border: 0.121/5.097/13.239 (5.334)
biter_area: 2.918/4.745/7.949 (4.626)

Disable logs in event_core.lua (master)

all: 0.023/1.852/12.854 (1.381)
river: 0.193/1.501/3.401 (1.353)
ordinary: 0.023/1.017/9.612 (0.124)
biter_area_border: 0.157/4.843/12.854 (4.817)
biter_area: 2.734/3.814/5.053 (3.777)

all: 0.022/1.926/9.706 (1.405)
river: 0.272/2.069/3.865 (1.912)
ordinary: 0.022/0.945/9.499 (0.119)
biter_area_border: 0.145/4.942/9.706 (5.087)
biter_area: 2.685/4.528/9.171 (4.472)
```

Conclusion: performance regression, ~+10% of run time. If I comment out performance logging in `event_core.lua`, goes back to normal as it was stated in description

Unfortunately, I've lost seed which I used to tests spawn area generation time and different seeds give you different time :(

morganchristiansson commented:
Maybe on_entity_clone event spam is causing it to be a bit slower? Can be called a lot in a benchmark.

DaTa- commented:
> Maybe on_entity_clone event spam is causing it to be a bit slower? Can be called a lot in a benchmark.

This is a benchmark of purely chunk generation, without chunk cloning included. It doesn't represent overall effect of disabling event performance logging for the scenario as a whole, just gives a hint than it might have performance impact, which should be investigated separately. There are no events should be fired during it (unless I've mistaken). The only thing I disabled were `log` calls, not performance collection itself.
I've tried to move `on_entity_cloned` out from event framework (just like `on_entity_damaged`), and only when I additionally replaced global variable lookup inside the callback with local cached version it gave like -5% performance gain for chunk cloning in total which was really hard to see behind my system background noise. So I suspect buffered io on my system (Windows 11) is bad or something
